### PR TITLE
exit if bundle bin path contains a path separator

### LINKED ISSUE #5485: Bundler does not work in paths that contain ":" (the colon character)
Not sure if the culprit is bundler or rubygems, but:

```console
$ bundle -v
Bundler version 1.14.5
$ pwd
/var/www/mlforums/versions/2017-02-28T23:35:32.512069Z
$ cat .bundle/config
---
BUNDLE_FROZEN: "1"
BUNDLE_PATH: "vendor/bundle"
BUNDLE_JOBS: "2"
BUNDLE_WITHOUT: "development:test"
BUNDLE_DISABLE_SHARED_GEMS: "true"
$ bundle exec gem env
RubyGems Environment:
  - RUBYGEMS VERSION: 2.6.8
  - RUBY VERSION: 2.4.0 (2016-12-24 patchlevel 0) [x86_64-linux]
  - INSTALLATION DIRECTORY: /var/www/mlforums/versions/2017-02-28T23:35:32.512069Z/vendor/bundle/ruby/2.4.0
  - USER INSTALLATION DIRECTORY: /home/mlforums/.gem/ruby/2.4.0
  - RUBY EXECUTABLE: /home/mlforums/.rbenv/versions/2.4.0/bin/ruby
  - EXECUTABLE DIRECTORY: /var/www/mlforums/versions/2017-02-28T23:35:32.512069Z/vendor/bundle/ruby/2.4.0/bin
  - SPEC CACHE DIRECTORY: /home/mlforums/.gem/specs
  - SYSTEM CONFIGURATION DIRECTORY: /home/mlforums/.rbenv/versions/2.4.0/etc
  - RUBYGEMS PLATFORMS:
    - ruby
    - x86_64-linux
  - GEM PATHS:
     - /var/www/mlforums/versions/2017-02-28T23:35:32.512069Z/vendor/bundle/ruby/2.4.0
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => false
     - :bulk_threshold => 1000
  - REMOTE SOURCES:
     - https://rubygems.org/
  - SHELL PATH:
     - /var/www/mlforums/versions/2017-02-28T23
     - 35
     - 32.512069Z/vendor/bundle/ruby/2.4.0/bin
     - /var/www/mlforums/versions/2017-02-28T23
     - 35
     - 32.512069Z/vendor/bundle/ruby/2.4.0/bin
     - /home/mlforums/.rbenv/versions/2.4.0/bin
     - /home/mlforums/.rbenv/libexec
     - /home/mlforums/.rbenv/plugins/ruby-build/bin
     - /home/mlforums/.rbenv/shims
     - /home/mlforums/.rbenv/bin
     - /usr/local/sbin
     - /usr/local/bin
     - /usr/sbin
     - /usr/bin
     - /sbin
     - /bin
     - /usr/games
     - /usr/local/games
```

Note the `SHELL PATH` above.

`bundle exec` then fails:

```console
$ bundle install -j2 --path vendor/bundle --deployment --without development test
$ bundle exec puma
bundler: command not found: puma
Install missing gem executables with `bundle install`
```

If I rename the directory to not contain `:`, `bundle exec puma` succeeds

### PR DESCRIPTION
**Pull request by ajwann**
fixes #5485

### What was the end-user problem that led to this PR?

Bundler does not work with paths that contain the path separator.

### What was your diagnosis of the problem?

The ```bundle_bin``` path can't contain a path separator.

### What is your fix for the problem, implemented in this PR?

We can validate the ```bundle_path``` to make sure that it does not contain the
current ```File::PATH_SEPARATOR```. If we find the path separator, we exit with a non-zero
code and give the user an error message. 

### Why did you choose this fix out of the possible options?

I chose this fix because it was recommended by @colby-swandale.

### CODE REVIEW

File: spec/bundler/shared_helpers_spec.rb [hunk 1]
```diff
@@ -261,6 +261,30 @@
...
+      begin
+        expect(subject.send(:validate_bundle_path)).to raise_exception(SystemExit).
+          and output("WARNING: Your bundle path contains a ':', which is the" \
+                     " path separator for your system. Please change your" \
+                     " bundle path path to not include ':'.")
+      rescue SystemExit => e
+        expect(e.status).to eq(1)
+      end
+
+      stub_const("File::PATH_SEPARATOR", "^".freeze)
```
Comment (colby-swandale) [inline spec/bundler/shared_helpers_spec.rb]: Is there a particular reason why we need this test as well?

File: spec/bundler/shared_helpers_spec.rb [hunk 2]
```diff
@@ -261,6 +261,30 @@
       subject.set_bundle_environment
     end
 
+    it "exits if bundle path contains the path seperator" do
```
Comment (colby-swandale) [inline spec/bundler/shared_helpers_spec.rb]: You have 2 distinct tests here, please each test one in their own `it` block.

File: spec/bundler/shared_helpers_spec.rb (hunk 1 above)
Comment (ajwann) [inline spec/bundler/shared_helpers_spec.rb]: I only added it to test that the error message changes when the path separator is different than ```:```. I suppose it's redundant. I can remove it.

**Comment by ajwann** [comment]
@colby-swandale I've made the requested changes.

File: lib/bundler/shared_helpers.rb [hunk 3]
```diff
@@ -203,6 +203,17 @@ def md5_available?
 
   private
 
+    def validate_bundle_path
+      return unless Bundler.bundle_path.to_s.match(File::PATH_SEPARATOR)
+      message = "WARNING: Your bundle path contains a '#{File::PATH_SEPARATOR}', " \
```
Comment (colby-swandale) [inline lib/bundler/shared_helpers.rb]: Remove 'WARNING:` as this is an error that will exit Bundler. We use warning when we need to notify the user about something wrong but Bundler can still continue with the action.

File: lib/bundler/shared_helpers.rb [hunk 4]
```diff
@@ -203,6 +203,17 @@ def md5_available?
...
+                "which is the path separator for your system. Bundler cannot " \
+                "function correctly when the Bundle path contains the " \
+                "system's PATH separator. Please change your " \
+                "bundle path to not include '#{File::PATH_SEPARATOR}'." \
+                "\nYour current bundle path is '#{Bundler.bundle_path}'."
+      raise Bundler::BundlerError, message
```
Comment (colby-swandale) [inline lib/bundler/shared_helpers.rb]: Can you raise `PathError` here instead (and in the test), this sets the correct error status, This was a mistake on my part.

**Comment by ajwann** [comment]
@colby-swandale thank you for your prompt feedback. I've made the requested changes.

File: lib/bundler/shared_helpers.rb [hunk 5]
```diff
@@ -203,6 +203,17 @@ def md5_available?
 
   private
 
+    def validate_bundle_path
+      return unless Bundler.bundle_path.to_s.match(File::PATH_SEPARATOR)
```
Comment (segiddins) [inline lib/bundler/shared_helpers.rb]: this can just use `include?(File::PATH_SEPARATOR)`

Comment (ajwann) [inline lib/bundler/shared_helpers.rb]: Done!

**Comment by segiddins** [comment]
@bundlerbot r+

**Pull request merged**